src/app/settings/components/project-role-policies-edit/project-role-policies-edit.tsx

```diff
@@ -0,0 +1,181 @@
+    if (action === undefined) {
+        action = '';
+    }
+    if (object === undefined) {
+        object = '';
+    }
+    if (permission === undefined) {
+        permission = '';
+    }
+    return `p,proj:${project}:${role},applications,${action},${project}/${object},${permission}`;
+}
+
+const actions = ['get', 'create', 'update', 'delete', 'sync', 'rollback', 'terminateop'];
+
+export const ProjectRolePoliciesEdit = (props: ProjectRolePoliciesProps) => (
+    <DataLoader load={() => services.applications.list([props.projName])} ref={(loader) => this.loader = loader}>
+    {(applications) => (
+        <React.Fragment>
+            <h4>Policies</h4>
+            { props.policies.length > 0 ? (
+                <div className='argo-table-list'>
+                    <div className='argo-table-list__head'>
+                        <div className='row'>
+                            <div className='columns small-4'>ACTION</div>
+                            <div className='columns small-4'>OBJECT</div>
+                            <div className='columns small-4'>PERMISSION</div>
+                        </div>
+                    </div>
+
+                    <div className='argo-table-list__row'>
+                    {props.policies.map((policy, i) => (
+                            <Policy key={i} field={['policies', i]}
+                                formApi={props.formApi} policy={policy}
+                                projName={props.projName}
+                                roleName={props.roleName}
+                                deletePolicy={() => props.formApi.setValue('policies', removeEl(props.policies, i))}
+                                availableApps={applications}
+                                actions={actions}
+                            />
+                            ))}
+                    </div>
+                </div>
+            ) : <div className='white-box'><p>Role has no policies</p></div> }
+            {!props.newRole && <a onClick={() => {
+                const newPolicy = generatePolicy(props.projName, props.roleName);
+                props.formApi.setValue('policies', (props.formApi.values.policies || []).concat(newPolicy));
+            }}>Add policy</a>}
+        </React.Fragment>
+    )}
+    </DataLoader>
+);
+
+interface PolicyProps {
+    projName: string;
+    roleName: string;
+    policy: string;
+    fieldApi: ReactForm.FieldApi;
+    actions: string[];
+    availableApps: Application[];
+    deletePolicy: () => void;
+}
+
+function removeEl(items: any[], index: number) {
+    items.splice(index, 1);
+    return items;
+}
+
+class PolicyWrapper extends React.Component<PolicyProps, any> {
+
+    public render() {
+        return (
+                <div className='row'>
+                    <div className='columns small-4'>
+                        <datalist id='action'>
```

Comment:
Safari does not support <datalist> . Please check if we can use polyfill to fix it. I think it is ok to do it in another PR later